helix-view/src/document.rs

```diff
@@ -633,6 +634,88 @@ impl Document {
         Some(fut.boxed())
     }
 
+    pub fn code_actions_on_save(
+        &self,
+    ) -> Option<BoxFuture<'static, Result<Vec<helix_lsp::lsp::CodeActionOrCommand>, anyhow::Error>>>
+    {
+        let code_actions_on_save = self
+            .language_config()
+            .map(|c| c.code_actions_on_save.clone())?;
+
+        if code_actions_on_save.is_empty() {
+            return None;
+        }
+
+        let request = self.code_actions(self.full_range())?;
+
+        let fut = async move {
+            log::debug!("Configured code actions on save {:?}", code_actions_on_save);
+            let json = request.await?;
+            let response: Option<helix_lsp::lsp::CodeActionResponse> =
+                serde_json::from_value(json)?;
+            let mut code_actions = match response {
+                Some(value) => value,
+                None => helix_lsp::lsp::CodeActionResponse::default(),
+            };
+            log::debug!("Available code actions {:?}", code_actions);
+            code_actions.retain(|action| {
+                matches!(
+                    action,
+                    helix_lsp::lsp::CodeActionOrCommand::CodeAction(x) if x.disabled.is_none() &&
+                        code_actions_on_save.iter().any(|a| match &x.kind {
+                            Some(kind) => kind.as_str() == a,
+                            None => false
+                    })
+                )
+            });
+            if code_actions.len() < code_actions_on_save.len() {
+                code_actions_on_save.iter().for_each(|configured_action| {
+                    if !code_actions.iter().any(|action| match action {
+                        helix_lsp::lsp::CodeActionOrCommand::CodeAction(x) => match &x.kind {
+                            Some(kind) => kind.as_str() == configured_action,
+                            None => false,
+                        },
+                        _ => false,
+                    }) {
+                        log::error!(
+                            "Configured code action on save is invalid {:?}",
+                            configured_action
+                        );
+                    }
+                })
+            }
```

Comment:
I don't think these code actions are necessarily returned on every import so this would potentially generate a lot of errors. This should atlest be an info` log but this also `O(NM)` so I am not sure we really need this anyway? 

Also, I think a for loop is basically always more idiomatic than `for_each`

Reply:
I agree this isn't great. I struggled to find a good solution for this. Ideally this would be done at the same time as the filtering with the `code_actions.retain()` above. Perhaps a `for` loop there instead? I can play around with it some more and see if I can come up with something. Otherwise I'm fine with leaving this out. It's a nice to have, but like you said perhaps not needed.